src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -2086,56 +2129,58 @@ private MemberDeclarationSyntax ParseMemberDeclarationOrStatementCore(SyntaxKind
                     return this.ParseTypeDeclaration(attributes, modifiers);
                 }
 
-                if (acceptStatement &&
-                    this.CurrentToken.Kind != SyntaxKind.CloseBraceToken &&
-                    this.CurrentToken.Kind != SyntaxKind.EndOfFileToken &&
-                    this.IsPossibleStatement(acceptAccessibilityMods: true))
-                {
-                    var saveTerm = _termState;
-                    _termState |= TerminatorState.IsPossibleStatementStartOrStop; // partial statements can abort if a new statement starts
-
-                    // Any expression is allowed, not just expression statements:
-                    var statement = this.ParseStatementNoDeclaration(allowAnyExpression: true);
-
-                    _termState = saveTerm;
-                    if (statement != null)
-                    {
-                        return _syntaxFactory.GlobalStatement(statement);
-                    }
-                }
-
-                // Everything that's left -- methods, fields, properties, 
-                // indexers, and non-conversion operators -- starts with a type 
-                // (possibly void). Parse that.
                 TypeSyntax type = ParseReturnType();
 
                 var afterTypeResetPoint = this.GetResetPoint();
 
                 try
                 {
-                    var sawRef = type.Kind == SyntaxKind.RefType;
+                    if (!haveAttributes && !haveModifiers && (type.Kind == SyntaxKind.RefType || !IsOperatorKeyword()))
+                    {
+                        this.Reset(ref startPoint);
+
+                        if (this.CurrentToken.Kind != SyntaxKind.CloseBraceToken &&
+                            this.CurrentToken.Kind != SyntaxKind.EndOfFileToken &&
+                            this.IsPossibleStatement(acceptAccessibilityMods: true))
+                        {
+                            var saveTerm = _termState;
+                            _termState |= TerminatorState.IsPossibleStatementStartOrStop; // partial statements can abort if a new statement starts
+                            bool wasInAsync = IsInAsync;
```

Comment:
Is this method only supposed to be called to parse things contained directly in the compilation unit? I'm wondering if what scenario do we get here and IsInAsync was already 'true'.

Reply:
> Is this method only supposed to be called to parse things contained directly in the compilation unit? I'm wondering if what scenario do we get here and IsInAsync was already 'true'.

I prefer to not make any assumptions, this is more robust in the long term.

---
In reply to: [382864944](https://github.com/dotnet/roslyn/pull/41706#discussion_r382864944) [](ancestors = 382864944)